w3d_translator/read_xml.py

```diff
@@ -1,50 +1,182 @@
 
-    story = {}  # TODO: Make story its own class (5)
+    # Parse each <PlacementRoot>, each <Placement> is referenced by name
+    story["walls"] = dict(
+        (tag.attrib.pop("name"), parse_recursive(tag))
+        for tag in root.find("PlacementRoot")
+    )
 
     # TODO: Build each <Object> in <ObjectRoot> (6)
-    # object_root = {}
-    for tag in root.find("ObjectRoot") or []:
-        pass  # Dict of objects by name
-
     # TODO: Build each <Group> in <GroupRoot> (7)
-    # group_root = {}
-    for tag in root.find("GroupRoot") or []:
-        pass  # Dict of groups by name (group -> array of object names)
 
+    # TODO: Build each <Sound> in <SoundRoot> (10)
+    # TODO: Build each <ParticleActionList> in <ParticleActionRoot> (11)
     # TODO: Build each <Timeline> in <TimelineRoot> (8)
-    # timeline_root = {}
-    for tag in root.find("TimelineRoot") or []:
-        pass  # Dict of Timeline by name
+    return story
 
-    # TODO: Build each <Placement> in <PlacementRoot> (9)
-    # placement_root = {}
-    for tag in root.find("PlacementRoot") or []:
-        pass  # Dict of Placement by name
 
-    # TODO: Build each <Sound> in <SoundRoot> (10)
-    # sound_root = {}
-    for tag in root.find("SoundRoot") or []:
-        pass  # Dict of Sound by name
+def parse_string(string: str) -> Union[bool, int, float, tuple, Path, str]:
+    """Parses the variable type of a given string
 
-    # TODO: Build each <ParticleActionList> in <ParticleActionRoot> (11)
-    # particle_action_root = {}
-    for tag in root.find("ParticleActionRoot") or []:
-        pass  # Dict of ParticleActionList by name
+    Args:
+        string (str): Input string
 
-    # Globals
-    g = root.find("Global")
-    # TODO: Build <CameraPos> and <CaveCameraPos> (12)
-    # story["camera_pos"] = parse_camera_pos(g.find("CameraPos"))
-    # story["cave_camera_pos"] = parse_cave_camera_pos(g.find("CaveCameraPos"))
-    story["background"] = Color(g.find("Background"))
-    story["wand_navigation"] = WandNavigation(g.find("WandNavigation"))
+    Returns:
+        Union[bool, int, float, tuple, Path, str]: Data parsed as correct type
 
-    return story
+    # Test boolean
+    >>> parse_string('True')
+    True
+    >>> parse_string('TRUE')
+    True
+    >>> parse_string('false')
+    False
+    >>> parse_string('FaLSe') # Bad capitalization passes
+    False
+
+    # Test integer
+    >>> parse_string('123')
+    123
+    >>> parse_string('-123')
+    -123
+
+    # Test float
+    >>> parse_string('123.4')
+    123.4
+    >>> parse_string('-123.4')
+    -123.4
+    >>> parse_string('b123.4') # Invalid, return string
+    'b123.4'
+
+    # Test tuple of integers
+    >>> parse_string('(1, 2, 3)')
+    (1, 2, 3)
+    >>> parse_string('(12, 2, -1)')
+    (12, 2, -1)
+
+    # Test tuple of floats
+    >>> parse_string('(1.1, 2.2, 3.3)')
+    (1.1, 2.2, 3.3)
+    >>> parse_string('(13.5, 2.5, -1.5)')
+    (13.5, 2.5, -1.5)
+
+    # Test paths
```

Comment:
How about this for the path testing?

```suggestion
    # Test Paths
    >>> p1 = parse_string('/folder/file.xml')
    >>> isinstance(p1, Path), p1.parts
    (True, ('/', 'folder', 'file.xml'))
    >>> p2 = parse_string('./folder/file.xml')
    >>> isinstance(p2, Path), p2.parts
    (True, ('folder', 'file.xml'))
    >>> p3 = parse_string('../folder/file.xml')
    >>> isinstance(p3, Path), p3.parts
    (True, ('..', 'folder', 'file.xml'))
    >>> p4 = parse_string('/folder/.')
    >>> isinstance(p4, Path), p4.parts
    (True, ('/', 'folder'))
    >>> p5 = parse_string('/folder/') # TODO: This is failing
    >>> isinstance(p5, Path), p5.parts
    (True, ('/', 'folder'))
    >>> p6 = parse_string('/folder')
    >>> isinstance(p6, Path), p6.parts
    (True, ('/', 'folder'))
    >>> p7 = parse_string('folder/file.xml') #TODO: 28
    >>> isinstance(p7, Path), p7.parts
    (True, ('folder', 'file.xml'))

```

Reply:
(This also deals with TODO 28. I think 28 looks like it's working at present.)